File: exercises/scala/002_adapted-response-pattern/src/main/scala/dev/fResult/akka/barista/BaristaActor.scala
package dev.fResult.akka.barista

import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
import akka.actor.typed.{ActorRef, Behavior}
import dev.fResult.akka.Coffee
import dev.fResult.akka.barista.BaristaCommand.{CoffeeReady, OrderCoffee}
import dev.fResult.akka.coffeeMachine.CoffeeMachineCommand.{BrewCoffee, PickupCoffee}
import dev.fResult.akka.coffeeMachine.{CoffeeMachineActor, CoffeeMachineCommand}

object BaristaActor:
  private final case class State(orders: Map[String, Coffee] = Map.empty)

  def apply(): Behavior[BaristaCommand] = Behaviors.setup { context =>
    val coffeeMachineActorRef = spawnCoffeeMachineActor(context)

    handleCommands(context, coffeeMachineActorRef, State())
  }

  private def handleCommands(context: ActorContext[BaristaCommand],
                             coffeeMachineActorRef: ActorRef[CoffeeMachineCommand],
                             state: State,
                            ): Behavior[BaristaCommand] = Behaviors.receiveMessage {

    case cmd@OrderCoffee(whom, coffee) =>
      onOrderCoffee(cmd, context, coffeeMachineActorRef, state)

    case CoffeeReady(coffee) =>
      onCoffeeReady(context, coffeeMachineActorRef, state)
  }

  private def onOrderCoffee(command: OrderCoffee,
                            context: ActorContext[BaristaCommand],
                            coffeeMachineActor: ActorRef[CoffeeMachineCommand],
                            state: State,
                           ): Behavior[BaristaCommand] = {

    val updatedState = state.copy(state.orders + (command.whom -> command.coffee))
    context.log.info(s"Barista: Orders${printOrder(updatedState.orders.toSet)}")

    coffeeMachineActor ! BrewCoffee(command.coffee, context.self)
    handleCommands(context, coffeeMachineActor, updatedState)
  }

  private def onCoffeeReady(context: ActorContext[BaristaCommand],
                            coffeeMachineActorRef: ActorRef[CoffeeMachineCommand],
                            state: State,
                           ): Behavior[BaristaCommand] = {

    coffeeMachineActorRef ! PickupCoffee
    handleCommands(context, coffeeMachineActorRef, state)
  }

  private def spawnCoffeeMachineActor(context: ActorContext[BaristaCommand]): ActorRef[CoffeeMachineCommand] =
    context.spawn(CoffeeMachineActor(), "coffee-machine")

  private def printOrder(orderSet: Set[(String, Coffee)]): String = {
    val formattedOrders = orderSet.map(order => s"${order._1}->${order._2}")

    s"[${formattedOrders.mkString(", ")}]"
  }
end BaristaActor

File: exercises/java/003_timers/src/main/java/com/akka/training/barista/BaristaActor.java
package com.akka.training.barista;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;
import com.akka.training.Coffee;
import com.akka.training.coffeeMachine.CoffeeMachineActor;
import com.akka.training.coffeeMachine.CoffeeMachineCommand;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class BaristaActor extends AbstractBehavior<BaristaCommand> {
  // Orders <Whom, Coffee>
  private final Map<String, Coffee> orders = new HashMap<>();
  // reference to the coffee-machine child actor, allowing to send messages to coffee machine
  private final ActorRef<CoffeeMachineCommand> coffeeMachine;
  private final ActorRef<CoffeeMachineCommand.CoffeeReady> coffeeMachineMessageAdapter;

  private BaristaActor(ActorContext<BaristaCommand> context) {
    super(context);
    // We spawn the CoffeeMachine as child actor in the private constructor where we have access to
    // both context and class fields.
    // It returns an ActorRef that we need to keep in the state, we will need it to interact with
    // the CoffeeMachine.
    coffeeMachine = context.spawn(CoffeeMachineActor.create(), "coffee-machine");

    coffeeMachineMessageAdapter =
        context.messageAdapter(
            CoffeeMachineCommand.CoffeeReady.class,
            BaristaCommand.WrappedCoffeeMachineCoffeeReady::new);
  }

  public static Behavior<BaristaCommand> create() {
    return Behaviors.setup(BaristaActor::new);
  }

  // Format the orders into expected format [whom1->coffee1,whom2->coffee2]
  public static String printOrders(Set<Map.Entry<String, Coffee>> orders) {
    return orders.stream()
        .map(kv -> String.format("%s->%s", kv.getKey(), kv.getValue()))
        .reduce((acc, s) -> acc + "," + s)
        .map(s -> "[" + s + "]")
        .orElse("[" + "]");
  }

  @Override
  public Receive<BaristaCommand> createReceive() {
    return newReceiveBuilder()
        .onMessage(BaristaCommand.OrderCoffee.class, this::onOrderCoffee)
        .onMessage(
            BaristaCommand.WrappedCoffeeMachineCoffeeReady.class,
            this::onWrappedCoffeeMachineCoffeeReady)
        .build();
  }

  private Behavior<BaristaCommand> onOrderCoffee(BaristaCommand.OrderCoffee command) {
    orders.put(command.whom(), command.coffee());

    getContext().getLog().info("Orders:{}", printOrders(orders.entrySet()));

    coffeeMachine.tell(
        new CoffeeMachineCommand.BrewCoffee(command.coffee(), coffeeMachineMessageAdapter));

    return this;
  }

  private Behavior<BaristaCommand> onWrappedCoffeeMachineCoffeeReady(
      BaristaCommand.WrappedCoffeeMachineCoffeeReady wrappedCoffeeReady) {

    CoffeeMachineCommand.CoffeeReady coffeeReady = wrappedCoffeeReady.coffeeReady();

    getContext().getLog().info("Barista: Picking up {}", coffeeReady.coffee());

    coffeeMachine.tell(new CoffeeMachineCommand.PickupCoffee());

    return this;
  }
}
